internal/apidiff/apidiff.go

```diff
@@ -0,0 +1,287 @@
+		if err != nil {
+			d.errs[imp] = err
+			continue
+		}
+
+		// Diff the current checked out change against remote HEAD base.
+		out, err := d.diff(modDir, imp, pkg, base)
+		if err != nil {
+			d.errs[imp] = err
+			continue
+		}
+
+		if out != "" && out != ignored {
+			d.results[imp] = out
+		}
+	}
+
+	return nil
+}
+
+func (d *differ) writeBase(baseModDir, imp, pkg string) (string, error) {
+	if err := cd(baseModDir); err != nil {
+		return "", err
+	}
+
+	base := path.Join(baseModDir, "pkg.master")
+	out, err := execE("apidiff", "-w", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+
+	// If there was an issue with loading a submodule, change into that
+	// submodule directory and try again.
+	if isSubModErr(out) {
+		parent := d.manualParent(imp)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err := execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+	return base, nil
+}
+
+func (d *differ) diff(modDir, imp, pkg, base string) (string, error) {
+	if err := cd(modDir); err != nil {
+		return "", err
+	}
+	out, err := execE("apidiff", "-incompatible", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+	if isSubModErr(out) {
+		parent := d.manualParent(imp)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err = execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+
+	return out, err
+}
+
+func (d *differ) loadManifest() error {
+	f, err := os.Open(repoMetadataPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return json.NewDecoder(f).Decode(&d.manifest)
+}
+
+func (d *differ) manualParent(imp string) string {
+	pkg := strings.TrimPrefix(imp, rootMod)
+	split := strings.Split(pkg, "/")
+
+	mod := rootMod
+	for _, seg := range split {
+		mod = path.Join(mod, seg)
+		if parent, ok := d.manifest[mod]; ok && parent.ClientLibraryType == "manual" {
+			return strings.TrimPrefix(mod, rootMod+"/")
+		}
+	}
+
+	return pkg
+}
+
+func shouldRun() (bool, error) {
+	head, err := execE("git", "log", "-1")
+	if err != nil {
+		return false, err
+	}
+	return !strings.Contains(head, "BREAKING_CHANGE"), nil
+}
+
+func cloneGoCloud() (string, error) {
+	temp, err := ioutil.TempDir("/tmp", "google-cloud-go-*")
+	if err != nil {
+		return "", err
+	}
+	_, err = execE("git", "clone", "https://github.com/googleapis/google-cloud-go", temp)
+	return temp, err
+}
+
+func isSubModErr(msg string) bool {
+	return strings.Contains(msg, "missing") || strings.Contains(msg, "required")
+}
+
+func cd(dir string) error {
+	if verbose {
+		log.Printf("+ cd %s\n", dir)
+	}
+	return os.Chdir(dir)
+}
+
+func execE(cmd string, args ...string) (string, error) {
```

Comment:
What does the E stand for?

Reply:
This is kind of a hold over from an earlier version - the E suffix means that it returns the error rather than panics (previously had a `mustExec` that panicked)

Reply:
I changed it to `exec` and alias `os/exec` to `osexec`